lib/Conversion/SeqToSV/FirRegLowering.cpp

```diff
@@ -20,34 +20,64 @@ using llvm::MapVector;
 
 #define DEBUG_TYPE "lower-seq-firreg"
 
-// Reimplemented from SliceAnalysis to use a worklist rather than recursion and
-// non-insert ordered set.
-static void
-getForwardSliceSimple(Operation *root,
-                      llvm::DenseSet<Operation *> &forwardSlice,
-                      llvm::function_ref<bool(Operation *)> filter = nullptr) {
-  SmallVector<Operation *> worklist({root});
-
-  while (!worklist.empty()) {
-    Operation *op = worklist.pop_back_val();
-
-    if (!op)
-      continue;
-
-    if (filter && !filter(op))
-      continue;
+bool ReachableMuxes::isMuxReachableFrom(seq::FirRegOp regOp,
+                                        comb::MuxOp muxOp) {
+  return llvm::any_of(regOp.getResult().getUsers(), [&](Operation *user) {
+    if (opBlocksReachability(user))
+      return false;
+    buildReachabilityFrom(user);
+    return reachableMuxes[user].contains(muxOp);
+  });
+}
 
-    for (Region &region : op->getRegions())
-      for (Block &block : region)
-        for (Operation &blockOp : block)
-          if (forwardSlice.insert(&blockOp).second)
-            worklist.push_back(&blockOp);
-    for (Value result : op->getResults())
-      for (Operation *userOp : result.getUsers())
-        if (forwardSlice.insert(userOp).second)
-          worklist.push_back(userOp);
-
-    forwardSlice.insert(op);
+void ReachableMuxes::buildReachabilityFrom(Operation *startNode) {
+  // This is a backward dataflow analysis.
+  // First build a graph rooted at the `startNode`. Every user of an operation
+  // that doesnot block the reachability is a child node. Then, the ops that
+  // are reachable from a node is computed as the union of the Reachability of
+  // all its child nodes.
+  // The dataflow can be expressed as, for all child in the Children(node)
+  // Reachability(node) = node + Union{Reachability(child)}
+  if (visited.find(startNode) != visited.end())
+    return;
+  // The stack to record enough information for an iterative post-order
+  // traversal.
+
+  llvm::SmallVector<OpUserInfo> stk;
+
+  stk.emplace_back(startNode);
+
+  while (!stk.empty()) {
+    auto &info = stk.back();
+    Operation *currentNode = info.op;
+
+    // Node is being visited for the first time.
+    if (info.userIter == info.userRange.begin())
+      visited.insert(currentNode);
```

Comment:
By inserting it here, seems like could put the same node multiple times on the visit stack -- haven't thought it through but would it make sense to update the set at point adding to the stack? (maybe the `visited.insert(child).second` pattern?

Reply:
A node is added to stack only if the iterator is not at the end of the range, that's why this check is here, to ensure the visited is populated even if the users of an op are empty. I don't think a node can be added multiple times to the stack.